Density-based Integral Screening and Incremental Fock Build

andyj10224 opened the pull request:
## Description
Implements Density matrix-based integral screening algorithms for Direct SCF, as well as Incremental Fock Build, the process of building a Fock matrix using the difference in the density matrix between SCF iterations.

Reference Paper:
https://onlinelibrary.wiley.com/doi/abs/10.1002/jcc.540100111

Updated Version of PR #2062

## Todos
<!-- Notable points (developer or user-interest) that this PR has or will accomplish. -->
- [x] Provide Massive Speedups to DirectJK Builds for HF and DFT
- [x] Accuracy of Calculations remain uncompromised

## Questions
- [x] Is my current method of handling the density matrix for UHF and ROHF ideal?

## Checklist
- [x] Passing cc-pVDZ and cc-pVTZ tests for HF and DFT (e_convergence 1.0e-6)
- [x] Passing aug-cc-pVDZ, aug-cc-pVTZ tests for HF and DFT

## Status
- [x] Ready for review
- [ ] Ready for merge

JonathonMisiewicz commented:
Marked as a 1.5 target.

Please fix any merge conflicts. I also recommend that "adding new database files" be moved to a separate PR.

andyj10224 commented:
Some graphs from running B3LYP on the S22 dimer set (Not S22x7, as it would take too long)


![DFT Eint Diff S22 IFB](https://user-images.githubusercontent.com/60559795/114731808-07c01b00-9d10-11eb-95fe-cff47eba9d69.png)
![DFT Eint Diff S22 DS](https://user-images.githubusercontent.com/60559795/114731810-07c01b00-9d10-11eb-9e4b-26689b8d2ad6.png)

![DFT Eabs Difference S22 IFB](https://user-images.githubusercontent.com/60559795/114731879-1575a080-9d10-11eb-970c-79c99f2b46eb.png)
![DFT Eabs Difference S22 DS](https://user-images.githubusercontent.com/60559795/114731811-0858b180-9d10-11eb-8d2c-ee5a87efdf34.png)

Calculation Wall Times (running the whole database, 8 cores):

Reference: 7186.40 seconds
Density Screening: 4126.07 seconds
Increfock: 3635.78 seconds

andyj10224 commented:
> Marked as a 1.5 target.
> 
> Please fix any merge conflicts. I also recommend that "adding new database files" be moved to a separate PR.

Will do, thank you!

hokru commented:
I am a bit confused what the graphs show, could you explain them a bit more?

JonathonMisiewicz commented:
Also, how have you verified correctness? While I'm not familiar with these methods, "can induce an energy difference of over 2 _hartrees_ even at conservative cutoffs" does not seem safe.

andyj10224 commented:
> I am a bit confused what the graphs show, could you explain them a bit more?

Yes, those graphs represent the distribution of differences expected from using density screening (or density screening + incremental Fock build) compared to not using density screening at all, on the total B3LYP energy of dimers, as well as the interaction energies in the S22 set (not S22x7). Results suggest that there is no harm in using density screening.

hokru commented:
What is the number in the top left corner (1e4, 1e-7) and what is plotted on the x-axis?

andyj10224 commented:
> Also, how have you verified correctness? While I'm not familiar with these methods, "can induce an energy difference of over 2 _hartrees_ even at conservative cutoffs" does not seem safe.

I have extensively verified the correctness of this approach, across many different molecules, basis sets, etc. I have never observed an error of "2 Hartrees". The worst I have ever observed is on the order of 1.0e-5 Hartrees, better than density fitting. Here is a link to my "benzene case study".

https://docs.google.com/spreadsheets/d/1HDMZ5PV6GhnK4i68Y1t86VhauTwqMP_X2kII21GCsiY/edit?usp=sharing

I empirically discovered that the best screening threshold given an energy convergence of 1.0e-n is 1.0e-(n+3) for non-augmented basis sets and 1.0e-(n+5) for augmented basis sets.

For example, for an e_convergence of 1.0e-8 in cc-pVTZ, the ideal threshold is 1.0e-11, while it is 1.0e-13 in aug-cc-pVTZ.

andyj10224 commented:
> What is the number in the top left corner (1e4, 1e-7) and what is plotted on the x-axis?

The numbers on the top left corner reference the order of magnitude (e.g. 1e-7 hartrees, or 1.0e-4 kcal/mol).

The values on the x-axis are violin plot represent distribution frequency (It's a violin plot).

JonathonMisiewicz commented:
Including the _order of magnitude_ in a corner of the plot is highly non-standard and leads to the confusion I just went through. Choose units that are of the proper order of magnitude, and include any needed decimals on the axis directly. According to your axes, you have 2 hartree error, which is utterly unacceptable. 0.2 microhartree error is much more reasonable.

What you're describing confirms that the error is small. That is different from confirming correctness. Are you able to compare the numbers to some other implementation and show agreement? Is there some rare property that the exact scheme has, which you can numerically reproduce?

andyj10224 commented:
> Including the _order of magnitude_ in a corner of the plot is highly non-standard and leads to the confusion I just went through. Choose units that are of the proper order of magnitude, and include any needed decimals on the axis directly. According to your axes, you have 2 hartree error, which is utterly unacceptable. 0.2 microhartree error is much more reasonable.
> 
> What you're describing confirms that the error is small. That is different from confirming correctness. Are you able to compare the numbers to some other implementation and show agreement? Is there some rare property that the exact scheme has, which you can numerically reproduce?

Ah, the energy convergence criteria itself is 1.0e-6, and the errors are well below that number, so I strongly believe that we are safe.

It may not be a good idea to compare to another implementation since every implementation has different tricks thrown in. Rather, if the energy difference is less than the e_convergence criteria, we are definitely safe. Though we could try to reproduce something like MBIS charges to check if we are safe.

hokru commented:
Now I also get it :)

Incremental Fock builds can show numerical error accumulation. Adding a reset frequency is sensible, e.g. after 20 incremental builds do a full Fock build from scratch. The actual number should be a (expert) user option.

The incremental Fock build can be compared against a standard build. They have to agree.
Comparing density screening to other programs can probably only be done for HF calculations and even then I would expect that one can only see 'similar errors' and no exact number agreement.

andyj10224 commented:
> Now I also get it :)
> 
> Incremental Fock builds can show numerical error accumulation. Adding a reset frequency is sensible, e.g. after 20 incremental builds do a full Fock build from scratch. The actual number should be a (expert) user option.
> 
> The incremental Fock build can be compared against a standard build. They have to agree.
> Comparing density screening to other programs can probably only be done for HF calculations and even then I would expect that one can only see 'similar errors' and no exact number agreement.

This is already done, as I turn off incremental Fock build after the density drops below 1.0e-5, to reduce the differences, though I could make this a user option. :)

JonathonMisiewicz commented:
Wait, what? Then how is Incremental Fock build set to energy tolerance on the order of 1e-9 giving you errors on the order of 1e-7?

andyj10224 commented:
> Wait, what? Then how is Incremental Fock build set to energy tolerance on the order of 1e-9 giving you errors on the order of 1e-7?

Oh no, I mean the SCREENING threshold is set to 1.0e-9, the energy tolerance is 1.0e-6.

JonathonMisiewicz commented:
Let me make sure I have this right: you're plotting energy errors on the order of 1e-7, but you only converged the energy to 1e-6? If so, then your plots are pure noise. If not, then what am I getting wrong?

andyj10224 commented:
> Let me make sure I have this right: you're plotting energy errors on the order of 1e-7, but you only converged the energy to 1e-6? If so, then your plots are pure noise. If not, then what am I getting wrong?

Yes sir that is right :)

JonathonMisiewicz commented:
You should not have made those graphs. All you know for sure is that the error is less than 1 micro hartree. You cannot tell the difference between 1.0 e-7 and 1.5 e-7, but people read graphs as if you can tell the difference between your data points.

This is also why you can't just present graphs. You need to be absolutely sure that your labels are clear, and that somebody who isn't you will be able to figure out what the graph means. If you need to add a sentence to explain them, do so.

I request additional benchmarks so we can get more precise estimates on how much error these techniques introduce. Can you increase energy convergence to `1e-10`?

Because this PR isn't coming in until 1.5 anyways, I'm going to turn my attention to other things for a while.

andyj10224 commented:
> You should not have made those graphs. All you know for sure is that the error is less than 1 micro hartree. You cannot tell the difference between 1.0 e-7 and 1.5 e-7, but people read graphs as if you can tell the difference between your data points.
> 
> This is also why you can't just present graphs. You need to be absolutely sure that your labels are clear, and that somebody who isn't you will be able to figure out what the graph means. If you need to add a sentence to explain them, do so.
> 
> I request additional benchmarks so we can get more precise estimates on how much error these techniques introduce. Can you increase energy convergence to `1e-10`?
> 
> Because this PR isn't coming in until 1.5 anyways, I'm going to turn my attention to other things for a while.

Ah, I am learning a lot. Good teachers like you will prepare me well for grad school :)

JonathonMisiewicz commented on psi4/driver/procrouting/scf_proc/scf_iterator.py:
```diff
@@ -452,6 +452,10 @@ def scf_iterate(self, e_conv=None, d_conv=None):
             self.Db().print_out()
 
+        # Set the current energy and density convergence values
+        core.set_variable("SCF DELTA E", Ediff)
```
Why add these two variables? They didn't exist in the code before this PR, and they aren't used elsewhere in this PR.

andyj10224 replied:
Ah, the DELTA D is used to determine when we can turn incremental fock build off, when we get close to convergence. Since I added DELTA D, I thought that it would make sense to add DELTA E as well. Personally, I do not like having to go to python to get delta D, and if you have any workarounds for this, I would love to hear it. :)

loriab replied:
Second removing `SCF DELTA E`. If we can't find another way to pass the rms density other than by global qcvar, I think it'd be better to `core.del_variable("SCF RMS D")` when it's no longer needed (after checking `has_variable`, I think).

JonathonMisiewicz commented on psi4/src/psi4/libfock/jk.cc:
```diff
@@ -540,6 +547,86 @@ void JK::AO2USO() {
+
+    if (iteration_ == 0 || D_ao_prev_.size() != D_ao_.size()) {
+
+        iteration_ = 0;
```
I am _strongly_ opposed to the JK object needing to know the iteration number. I'll have a more concrete counter-recommendation when I can read the original paper.

JonathonMisiewicz replied:
I'm still opposed to having an `iteration_` member variable, but as I look over the code, I don't understand it. Why are you treating D, J, K, and wK on the same footing in the `if` case, but treating D alone in the `else` case?

JonathonMisiewicz reviewed (changes requested):
I feel like I need a copy of the original Ahlrichs paper before I can review this PR - I don't understand the "big picture."

While I'm waiting for the library to get that to me, here are some initial comments.

susilehtola commented on psi4/src/read_options.cc:
```diff
@@ -1439,6 +1439,14 @@ int read_options(const std::string &name, Options &options, bool suppress_printi
+        /*- Threshold for SCF Density Screening -*/
+        options.add_double("DENSITY_SCREENING_TOLERANCE", 1.0e-13);
+        /*- Density Threshold to stop Incremental Fock Build -*/
+        options.add_double("IFB_D_CONVERGENCE", 1.0e-5);
```
This could just be `INCFOCK_CONVERGENCE`

JonathonMisiewicz replied:
Susi has a good point: "IFB" is too opaque. Please rename the option.

susilehtola commented:
Incremental Fock builds have the tendency to have numerical creep. As small changes incrementally accumulate in the density matrix, the incrementally formed Fock matrix may no longer be correct; this is especially noticeable in large systems. Typically this is solved by doing full Fock builds every `N` steps. How is this handled in your implementation?

JonathonMisiewicz commented on psi4/src/psi4/libmints/twobody.h:
```diff
@@ -144,6 +145,12 @@ class PSI_API TwoBodyAOInt {
+    /// Density Screening Tolerance
+    double density_screening_threshold_;
+    /// Max Density per Shell Pair
+    std::vector<std::vector<double>> max_dens_shell_pair_;
```
Move this line of code to be near `function_pair_values_` and its brethren: they're all giving lookup information used by the various screening tests. Similar to those functions, this vector of vectors should be flatten into a simple vector.

JonathonMisiewicz reviewed (changes requested):
Now that I have a copy of the paper, I've been able to review the density screening part of the PR. That half of the PR doesn't follow the standards of similar code, which is the big problem here, and fixing it requires a restructure. I request that you refactor the density screening part of this before I look at the incremental Fock build part. Trying to do both at once is too likely to cause confusion.

As for your handling of UHF and ROHF, I don't have expertise there. If you haven't already, do a quick search for density screening for open-shell references, and check the papers that cite Ahlrichs for any titles that specifically mention open-shells or new developments in Direct SCF.

andyj10224 commented:
> Now that I have a copy of the paper, I've been able to review the density screening part of the PR. That half of the PR doesn't follow the standards of similar code, which is the big problem here, and fixing it requires a restructure. I request that you refactor the density screening part of this before I look at the incremental Fock build part. Trying to do both at once is too likely to cause confusion.
> 
> As for your handling of UHF and ROHF, I don't have expertise there. If you haven't already, do a quick search for density screening for open-shell references, and check the papers that cite Ahlrichs for any titles that specifically mention open-shells or new developments in Direct SCF.

Hi Jonathon, I just implemented the refactoring. To request density screening, the user would just set `screening density` in the options

JonathonMisiewicz commented on tests/CMakeLists.txt:
```diff
@@ -41,13 +41,13 @@ foreach(test_name adc1 adc2 casscf-fzc-sp casscf-semi casscf-sa-sp ao-casscf-sp
-                  cc9 cc9a cdomp2-1 cdomp2-2 cepa1
-                  cepa2 cepa3 cepa-module ci-multi cisd-h2o+-0 cisd-h2o+-1
+                  cc9 cc9a cdomp2-1 cdomp2-2 cepa1 cepa2 cepa3 cepa-module 
+		  ci-multi cisd-h2o+-0 cisd-h2o+-1
```
Formatting.

JonathonMisiewicz commented on psi4/src/read_options.cc:
```diff
@@ -1439,6 +1439,14 @@ int read_options(const std::string &name, Options &options, bool suppress_printi
+        options.add_double("DENSITY_SCREENING_TOLERANCE", 1.0e-13);
+        /*- Density Threshold to stop Incremental Fock Build -*/
+        options.add_double("IFB_D_CONVERGENCE", 1.0e-5);
 
```
Incremental Fock build test cases still needed.

JonathonMisiewicz commented on psi4/src/read_options.cc:
```diff
@@ -1439,6 +1439,14 @@ int read_options(const std::string &name, Options &options, bool suppress_printi
+        options.add_double("DENSITY_SCREENING_TOLERANCE", 1.0e-13);
+        /*- Density Threshold to stop Incremental Fock Build -*/
+        options.add_double("IFB_D_CONVERGENCE", 1.0e-5);
 
```
User-facing documentation still needed.

JonathonMisiewicz commented on psi4/src/psi4/libmints/twobody.h:
```diff
@@ -198,6 +204,8 @@ class PSI_API TwoBodyAOInt {
     /*
      * Sieve information
      */
+    /// Update the Max Density Per Shell Pair given an updated Density Matrix (Haser 1989)
```
```suggestion
    /// Update max_dens_shell_pair_ given an updated density matrix (Haser 1989)
```

JonathonMisiewicz commented on psi4/src/psi4/libmints/twobody.cc:
```diff
@@ -112,6 +116,61 @@ TwoBodyAOInt::TwoBodyAOInt(const TwoBodyAOInt &rhs) : TwoBodyAOInt(rhs.integral_
+    
+    timer_on("Update Density");
+#pragma omp parallel for
+    for (int M = 0; M < nshell_; M++) {
```
Why are these loops independent? This code counts (M, N) and (N, M). I believe it would be more efficient to only count one.

JonathonMisiewicz commented on psi4/src/psi4/libmints/twobody.cc:
```diff
@@ -112,6 +116,61 @@ TwoBodyAOInt::TwoBodyAOInt(const TwoBodyAOInt &rhs) : TwoBodyAOInt(rhs.integral_
+    double Q_MN_sq = shell_pair_values_[N * nshell_ + M];
+    double Q_RS_sq = shell_pair_values_[S * nshell_ + R];
+
+    std::initializer_list<double> block_densities {max_dens_shell_pair_[M * nshell_ + N], max_dens_shell_pair_[R * nshell_ + S], 
```
This isn't what I meant when I asked for an initializer list. What I recommend was like [example](https://en.cppreference.com/w/cpp/algorithm/max) `std::max({1, 10, 50, 200})` is the example I had in mind. They don't need to use an initializer list class.

JonathonMisiewicz commented on psi4/src/psi4/libmints/twobody.cc:
```diff
@@ -112,6 +116,61 @@ TwoBodyAOInt::TwoBodyAOInt(const TwoBodyAOInt &rhs) : TwoBodyAOInt(rhs.integral_
+    max_density = max_density * max_density;
+
+    // Equations 6, 9, and 14
+    if (Q_MN_sq * Q_RS_sq * max_density > density_threshold) return true;
+    return false;
```
```suggestion
    return Q_MN_sq * Q_RS_sq * max_density > density_threshold;
```

JonathonMisiewicz commented on psi4/src/psi4/libmints/twobody.cc:
```diff
@@ -112,6 +116,61 @@ TwoBodyAOInt::TwoBodyAOInt(const TwoBodyAOInt &rhs) : TwoBodyAOInt(rhs.integral_
+    // Equation 6
+    double max_density = std::max(block_densities);
+    // Squared to account for the fact that Q_MN is given as its square
+    max_density = max_density * max_density;
```
Reading the paper, this is a major problem in your implementation. Haser's equation is that:
ε > Q_MN * Q_RS * ΔD.

You do not have his equation 13 correct. What you call Q is what Haser calls Q squared. To "correct" for this, you square ΔD, but then you would need to square ε as well. So when the user inputs a screening threshold, they are effectively getting the square root of that threshold. That's bad.

The simpler thing to do is just define Q as Haser does.

andyj10224 replied:
Ah, I set the density threshold to the square of the screening threshold, so my code was still valid, let me change it to make it more clear :) Schwarz Screening also does this, to avoid taking expensive square roots.

JonathonMisiewicz replied:
I'm going to request a second opinion from somebody more knowledgeable about screening procedures (@andysim?). What bothers me is that `screening_threshold_` is determined by `INTS_TOLERANCE`, and the documentation on `INTS_TOLERANCE` mentions nothing about this square/square-root behavior. At the very least, this is undocumented behavior.

JonathonMisiewicz commented on psi4/src/psi4/libmints/twobody.cc:
```diff
@@ -76,10 +76,14 @@ TwoBodyAOInt::TwoBodyAOInt(const IntegralFactory *intsfactory, int deriv)
     
     if (screening_threshold_ == 0.0) screening_type_ = ScreeningType::None;
+
+    max_dens_shell_pair_.resize(bs1_->nshell() * bs1_->nshell(), 1.0);
```
Why is this necessary?

andyj10224 replied:
Moved this to the `create_sieve_pair_info()` function, where its other brethren are set up too.

JonathonMisiewicz reviewed (changes requested):
I suspect the density screening algorithm is implemented incorrectly. I don't want to take focus away from this point, so I'm going to refrain from delving into the IFB just yet.

andyj10224 commented:
> I suspect the density screening algorithm is implemented incorrectly. I don't want to take focus away from this point, so I'm going to refrain from delving into the IFB just yet.

Hi Jonathan, I have added test cases for density screening and incfock in the pytests/test_erisieve.py file. I have yet to user-facing documentation, but feel free to review the code I have currently.

andyj10224 commented:
I just updated the docs too. One thing I am confused about, why is the `screening` keyword only in the sapt module?

hokru commented:
> One thing I am confused about, why is the `screening` keyword only in the sapt module? I think we should refactor it as a global module.

That's odd! 
Since we have `INTS_TOLERANCE` for every module maybe we should do the same with `screening`?

JonathonMisiewicz commented:
> I just updated the docs too. One thing I am confused about, why is the `screening` keyword only in the sapt module?

Very good question. @zachglick [added the option](https://github.com/psi4/psi4/commit/e1977af9ae69b77ca518c3169efb23018b51ffb0), so perhaps he can shed some light on this one.

loriab commented:
`SCREENING` should probably be a top-o-the-read_options-file keyword so that it distributes identically to the modules. keywords read by libraries are iffy to place since prepare_options_for_module(mod) but mod != lib. But I may be missing a trick as to why it was placed in sapt.

zachglick commented:
I can't remember or come up with a good reason for placing `SCREENING` in sapt. I agree that moving the option to the top of read_options makes a lot of sense.

JonathonMisiewicz commented on psi4/src/psi4/libfock/jk.h:
```diff
@@ -303,6 +314,21 @@ class PSI_API JK {
     std::vector<SharedMatrix> wK_ao_;
 
+    /// D, J, K, wK Matrices from previous iteration, used in Incremental Fock Builds
+    std::vector<SharedMatrix> D_ao_prev_;
```
Can you elaborate on why this needs to be a vector? For alpha and beta?

andyj10224 replied:
Ah, yes, D_ao_ is also a vector of matrix pointers, so D_ao_prev_ would just be an analog to it (indeed for alpha and beta in UHF and ROHF).

JonathonMisiewicz reviewed (changes requested):
Additional "miscellanea" items:

* I don't like the fact that a user who inputs X as their integral tolerance is actually getting SQRT(X). I want a second opinion from somebody more knowledgeable about the screening code. If nothing else, this needs to be stated very clearly in the documentation for the integral tolerance option.
* The "screening" keyword should be moved to top-level and have absolutely nothing to do with sapt. Check anywhere this keyword is used in the docs, and make sure the sapt prefix is fone.
* We need user-facing documentation on what an incremental Fock build is.
* I don't like that we have to add so many variables that are just for IFB, but I don't see a way around it. If other core developers have a recommendation here?
* I still want to remove knowledge of `iteration_` from JK. I need an answer to the question I asked about the `*_ao_prev_` variables before I can recommend how to do that.

andyj10224 commented on psi4/src/psi4/libfock/DirectJK.cc:
```diff
@@ -272,51 +272,62 @@ void DirectJK::compute_JK() {
+        if (density_screening_) ints[0]->update_density(D_ref);
         for (int thread = 1; thread < df_ints_num_threads_; thread++) {
             ints.push_back(std::shared_ptr<TwoBodyAOInt>(ints[0]->clone()));
+            if (density_screening_) ints[thread]->update_density(D_ref);
```
You are right, I added the density copying to the copy constructor for a TwoBodyAOInt

andyj10224 commented on CMakeLists.txt:
```diff
@@ -305,7 +305,7 @@ ExternalProject_Add(psi4-core
 
 add_subdirectory(external/downstream)
-add_subdirectory(doc)
+# add_subdirectory(doc)
```
Fixed it, some of the Sherrill machines didn't compile with this line, so I removed it temporarily

andyj10224 commented on psi4/src/export_mints.cc:
```diff
@@ -1231,18 +1232,23 @@ void export_mints(py::module& m) {
 
     typedef bool (TwoBodyAOInt::*compute_shell_significant)(int, int, int, int);
     typedef size_t (TwoBodyAOInt::*compute_shell_ints)(int, int, int, int);
+    typedef void (TwoBodyAOInt::*update_densities)(const std::vector<SharedMatrix>&);
```
It is not, it was there because I was trying to debug. Fixed.

andyj10224 commented on psi4/src/psi4/libfock/jk.h:
```diff
@@ -254,6 +254,17 @@ class PSI_API JK {
+    /// Perform Incremental Fock Build for J and K Matrices?
+    bool ifb_;
+    /// Density Matrix convergence criteria
+    double d_conv_;
```
Not used, removed.

andyj10224 commented on psi4/src/psi4/libfock/jk.h:
```diff
@@ -254,6 +254,17 @@ class PSI_API JK {
+    /// Density Matrix convergence criteria
+    double d_conv_;
+    /// The density matrix convergence value at which to stop performing IFB
+    double ifb_d_conv_;
```
Let's wait until we get other opinions on INCFOCK before we make any changes :)

andyj10224 commented:
I made the requested changes, I am wondering which .rst file I need to put the incfock documentation in. Waiting for expert opinions on incfock :)

loriab commented:
> I am wondering which .rst file I need to put the incfock documentation in.

I think https://github.com/psi4/psi4/blob/master/doc/sphinxman/source/scf.rst would be good.

JonathonMisiewicz commented:
> I made the requested changes, I am wondering which .rst file I need to put the incfock documentation in. Waiting for expert opinions on incfock :)

Waiting for more feedback sounds good to me. Libfock is one of the pieces of the code I don't know as well.

andyj10224 commented:
Hi all, 

I just updated the density screening code with the UHF/ROHF cases. I also added a test in the test_erisieve.py pytest file that ensures the same number of shell quartets are being screened out with RHF and UHF.

JonathonMisiewicz commented:
@andyj10224 @andysim What's the status of this PR? The last I heard was that on the 24th, Andy S. said, "i’m about to have a call with Andy so i’ll ask him some of the questions i have regarding density screening because i think it should be possible to do CSAM and density screening at the same time"

Are we still looking for another reviewer before re-considering my outstanding comments? Did something come of that conversation that is causing more development of this PR?

andyj10224 commented:
> @andyj10224 @andysim What's the status of this PR? The last I heard was that on the 24th, Andy S. said, "i’m about to have a call with Andy so i’ll ask him some of the questions i have regarding density screening because i think it should be possible to do CSAM and density screening at the same time"
> 
> Are we still looking for another reviewer before re-considering my outstanding comments? Did something come of that conversation that is causing more development of this PR?

Ah, @andysim was interested in doing density screening in addition to CSAM, not as its own separate screening. Other than that I think we are good to go on the PR. Do you have any comments for incremental Fock build? @andysim should I go ahead and make the change to do density screening in addition to things like CSAM?

JonathonMisiewicz commented:
Unless adding it will force you to rewrite a large amount of code, or not adding it will make your results incorrect, I'm going to request you leave that to a separate PR. It's difficult enough finding people to review this, and adding more code will only make it worse.

JonathonMisiewicz reviewed (changes requested):
The PR is coming along, but I suspect that I have been very confused about what this PR is for. If these changes were meant for the DirectJK algorithm only, yes, the title says that, but your description didn't, and the files you chose to edit led me to believe otherwise.

andyj10224 commented:
@JonathonMisiewicz To reference the treating the D separately, we are calculating the change in the density matrix, or delta D if we are not in the first iteration, the delta J, K, and wKs are added to the J and K in the post-iteration step. That way, incremental Fock build is available for all JK types. Basically knowledge of the iteration is needed so we know when to set up the delta D matrices. (It is also needed for my future CFMM and linK PRs, where we do NOT want to run linK or CFMM on the SAD guess)

andyj10224 commented:
One question, to all core developers, IMHO, we should change the default value for `df_scf_guess` from true to false, since optimized direct algorithms would be faster than density fitting for large systems. What do y'all think?

susilehtola commented:
> One question, to all core developers, IMHO, we should change the default value for `df_scf_guess` from true to false, since optimized direct algorithms would be faster than density fitting for large systems. What do y'all think?

But that's only true for small basis sets.

If you run small to moderate sized molecules with triple-zeta basis sets, DF is very fast.

JonathonMisiewicz commented:
> I made all of the requested changes, barring the `iteration_` variable, which I personally believe is necessary. If you have a workaround, please let me know :).This PR should be _really_ close to being finalized.

At risk of being grumpy, please don't say you made all the requested changes unless you actually have. There were many of my comments that still haven't been addressed.

susilehtola commented:
I would also consider turning off the incremental build once a preset level of `D_CONVERENCE` is achieved. The point about full rebuilds every `N` iterations was just that if, say, 40 SCF iterations are taken until the incremental SCF converges, it may have accumulated so much numerical creep that the actual orbital gradient is still significant.

andyj10224 commented:
> Thanks. The question of "when do we _use_ an incremental build?" needs to be settled (which it isn't yet - see Susi's comment) before I can approve this.

Hi Jonathon, I just implemented Susi's suggestion about the `d_convergence`. I also updated the necessary tests (`density-screen-1` and `density-screen-2`).

andyj10224 commented:
Besides the `Inc. Fock Build` printing issue, this PR is good to go, as I have just compiled and cleaned up the sphinxman docs.

<img width="796" alt="Density Screening Screenshot" src="https://user-images.githubusercontent.com/60559795/138020786-0dfba71c-08c1-44b1-bec1-ad5cf6a78005.png">

jturney reviewed (approved):
Super nice! Thanks for the feature!

andysim commented on doc/sphinxman/source/bibliography.rst:
```diff
@@ -558,3 +558,8 @@ Bibliography
+.. [Haser:1989:104]
+   M. Haser and R. Ahlrichs,
+   *J. Comp. Chem.* **10(1)**, 104 (1989).
+   doi: 10.1002/jcc.540100111
```
```suggestion
   doi: 10.1002/jcc.540100111
   
.. [Thompson:2017:144101]
   T. H. Thompson and C. Ochsenfeld
   *J. Chem. Phys.* **147**, 144101 (2017).
   doi: 10.1063/1.4994190
```

andysim commented on doc/sphinxman/source/scf.rst:
```diff
@@ -705,7 +705,29 @@ post SCF algorithms require a specific implementation.
+.. math:: (PQ|RS) <= \sqrt{(PQ|PQ)(RS|RS)}
+
+CSAM
+    An extension of the Schwarz estimate that also screens over the long range 1/r operator.
```
```suggestion
    An extension of the Schwarz estimate that also screens over the long range 1/r operator, described in [Thompson:2017:144101]_.
```

andysim commented on psi4/src/psi4/libfock/DirectJK.cc:
```diff
@@ -271,52 +363,83 @@ void DirectJK::compute_JK() {
         for (int thread = 0; thread < df_ints_num_threads_; thread++) {
             ints.push_back(std::shared_ptr<TwoBodyAOInt>(factory->erf_eri(omega_)));
+            if (density_screening_) ints[thread]->update_density(D_ref);
         }
```
If this shows up in the timings for parallel runs, we can change it to building a single integral object, then cloning it for all threads with an ID greater than 0.  It's correct as-is, so I don't suggest we change it unless we see L2 object construction times becoming significant.

andyj10224 replied:
iirc, update density is a REALLY cheap step.

andyj10224 replied:
`Update Density                      :      6.883u      0.050s      0.234w      9 calls`

Cheapness confirmed, this is on an 18 core machine.

andysim reviewed (approved):
This is really outstanding work - kudos for getting it all figured out, and implemented in some of the most difficult parts of the code in terms of control flow.  My concern before was that the list of max densities is an (N^2) quantity and could get large for larger systems, while storing it only for those shell pairs that survive the Schwarz/CSAM screening would not.  This is not really a valid concern because we have overlap matrices and other O(N^2) quantities in core anyway, so the implementation is great as-is.  I added a suggestion or two to add the  CSAM reference because I can never remember which paper it comes from.  LGTM!

JonathonMisiewicz commented:
Per Lori request, we have a hold on merging this in until David Poole can look this over.

davpoolechem commented on psi4/src/psi4/libfock/jk.h:
```diff
@@ -706,6 +706,34 @@ class PSI_API DirectJK : public JK {
+
+    // => Incremental Fock build variables <= //
+    /// Perform Incremental Fock Build for J and K Matrices?
+    bool incfock_;
```
Would it be better to change the name of this variable to "do_incfock_"? It seems that the do_ at the beginning of the variable name is something of a convention for variables like this (e.g., do_csam_, do_J_). Also, it would probably be wise to put in the comment what this variable defaults to.

andyj10224 replied:
Good idea, but at this point, changing the associated variable would be too much of a "shotgun surgery"

davpoolechem commented on psi4/driver/procrouting/scf_proc/scf_iterator.py:
```diff
@@ -425,6 +428,9 @@ def scf_iterate(self, e_conv=None, d_conv=None):
                 if self.frac_performed_:
                     status.append("FRAC")
 
+                if incfock_performed:
```
I notice that this if statement uses an incfock_performed variable determined via parameters in jk, while the if statements directly above this one use variables contained within the HF class (MOM_performed_ and frac_performed_). For consistency's sake, it might be better to move incfock_performed_ to the HF class and keep this if statement as "if self.infock_performed_:"

JonathonMisiewicz replied:
Absolutely not.

1) Class variables are a necessary evil at best. (Good luck trying to figure out where the variable was last modified.) Local variables are preferred.
2) `incfock_performed_` isn't something that the HF object should know about. It doesn't mean anything unless you're using a direct algorithm.

davpoolechem replied:
All right, that makes sense. I had mainly suggested it for consistency's sake; but if it is better this way, then there is no real need to change it.

davpoolechem commented on psi4/driver/procrouting/scf_proc/scf_iterator.py:
```diff
@@ -436,6 +442,7 @@ def scf_iterate(self, e_conv=None, d_conv=None):
         core.timer_off("HF: Form D")
 
         self.set_variable("SCF ITERATION ENERGY", SCFE)
+        core.set_variable("SCF D NORM", Dnorm)
```
It's interesting to me that the "SCF D NORM" variable is part of core rather than self. Is there any reason for this, and would it be better for SCF D NORM to be part of self here instead of core?

loriab replied:
Good catch -- it alarmed me, too, at one point. iirc, the answer was that directJK that reads this doesn't have access to the wfn object, so globals is the easiest way to communicate.

davpoolechem replied:
Ahhhhhhhh, I see. That would make sense!

davpoolechem commented on psi4/src/read_options.cc:
```diff
@@ -1439,6 +1439,11 @@ int read_options(const std::string &name, Options &options, bool suppress_printi
+        options.add_bool("INCFOCK", false);
+        /*- Frequency with which to compute the full Fock matrix if using |scf__incfock| . 
+        N means rebuild every N SCF iterations to avoid accumulating error from the incremental procedure. -*/
+        options.add_int("INCFOCK_FULL_FOCK_EVERY", 5);
```
Just as a heads up, I was able to run the IncFock code with values of INCFOCK_FULL_FOCK_EVERY below 1. Running negative values for this option seems to perform incfock for all iterations, while setting this option to 0 leads to a floating-point exception on my system. It may be wise to add some sort of option checking to make sure that the input value for this variable makes sense.

andyj10224 replied:
Just fixed

davpoolechem reviewed (commented):
I just finished looking at the code. I left a couple of comments to be reviewed before I approve these changes.

andyj10224 commented:
I addressed all of David's comments. Ready to merge.

JonathonMisiewicz commented:
I'd rather have @davpoolechem officially sign off on this first. It'll be an hour before the test suite check finishes, anyway.

davpoolechem reviewed (approved):
Andy and I talked a bit about the comments that I made, and he has sufficiently addressed all of them. I give my seal of approval to this code!
